Dockerise Script & Docker Improvements

joeShuff opened the pull request:
Hi, 

Saw your project on reddit and was impressed someone managed to do this, I thought of trying it myself when I switched for a day or two during the december peaks!

I've not managed to try your script out yet as my Home Mini is playing up but once its sorted I plan to run it!

_I saw your comment on the issue asking for a docker container so am open to your thoughts. Or if you'd rather not merge this into your repo I'll try to keep on top of any changes you make in my own fork._

# This PR

## Docker!
I've added docker files for building and the pipelines for automatically creating/publishing a build when you create a release.

If you merge this, you need to add 2 secrets to the repository, in Settings -> Secrets and Variables -> Actions and add
DOCKER_USERNAME and DOCKER_PASSWORD. 

## Changes to work with docker
Dockerisation has meant I had to change a few things. 

I've exposed the config via environment variables so there is no need to change python files. Then changed all uses of those in the script.

## Inbuilt Scheduling
CRON jobs can be scary for most, I'm a professional developer and still hate setting them up. I've added a simply wrapper python script that will take your desired time as a string HH:MM and keep looping until its that time of the day, then execute your script.

I've updated the readme with the new way of deploying.

Thanks!

eelmafia commented:
Hey,

Thanks for all the work! I'd be more than happy to merge this. I have only 2 concerns:

1. If the config file is no longer used for anything except the URL then it wouldn't be possible to manually run this (no Docker). 
2. Can you test just the Playwright function to make sure it works. https://github.com/eelmafia/octopus-minmax/issues/9#issuecomment-2629165728 noticed Playwright can fail silently if it's not set up properly. I haven't had time to test yet, but I wonder if using the Python Playwright image from the [docs ](https://playwright.dev/python/docs/docker)would be the best way to handle this.

joeShuff commented:
1. I understand your concern with that. We can take one of two approaches.
    - Do similar as in #6 where we support both environment variables and the config file.
    - Or (my preference cause I think its cleaner as a consumer) to support arguments for the run command. Then the manual run would be something like
      ```
       python your_script.py \
      --account_number "<your_account_number>" \
      --api_token "<your_api_token>" \
      --octopus_email "<your_email>" \
      --octopus_password "<your_password>" \
      --execution_time "23:00" \
      --discord_webhook "<your_webhook_url>"
      ```
My concern with the config file is it adds that "need to change a file" requirement which can turn a lot of non-technical people off. But no reason we can't have both ways.

2. I can give it a try, as I don't have the home mini working the script isn't functional for me right now, which is why I didn't change anything about it other than dockerising.

Side note: I can see the use case where people want to just fire a one-time run rather than using the built in scheduling. I will add support for passing a "one-time" variable to simply trigger the script and then exit.

joeShuff commented:
Just to add for anyone that is dying to run docker and cant wait for this PR, I did a test push to my docker account so the image is usable. Follow the readme of my fork to get it setup and give it a try. 

If you do try it and have any feedback because I've not actually ran the switch via docker, then let me know.

Image: denizenn/octopus-minmax-bot
My readme: https://github.com/joeShuff/octopus-minmax/blob/main/README.md

eelmafia commented:
1. I think the first approach is better IMO. Non-technical people would probably struggle with long terminal commands and editing a text file is simpler. 
2. You could test Playwright without the mini. Just run only `switch_tariff("Go")` and have Playwright save a screenshot instead of initiating the switch. 

```
def switch_tariff(target_tariff):
    with sync_playwright() as playwright:
        ### Add this
        try:
            browser = playwright.chromium.launch(
               headless=True)
        except Exception as e:
            print(e)  # Should print out if its not working 
        context = browser.new_context( viewport= {"width": 1920, "height": 1080} )
        page = context.new_page()
        page.goto("https://octopus.energy/")
        ...
        page.goto(f"https://octopus.energy/smart/{target_tariff.lower()}/sign-up/?accountNumber={config.ACC_NUMBER}")
        page.wait_for_timeout(10000)
        ### Add this
        page.screenshot(path="screenshot.png") 
        #page.locator("section").filter(has_text="Already have a SMETS2 or “").get_by_role("button").click()
        #page.wait_for_timeout(10000)
        #check if url has success
        context.close()
        browser.close()
```
---
Btw, since you mentioned your mini is acting up, have you tried keeping it next to the elec smart meter? My mini had issues when I first got it and the fix for me was to put it on the floor right under the meter itself.

Johnr24 commented:
I agree here, I consider myself less technical (I am not profiesionally in the industry or a Dev) than most and have a strong dislike of anything with excessive terminal --flags or commands I'd much rather edit a text file than do it from commandline where possible

joeShuff commented:
Updated the config to allow changing the config.py file and also use environment variables. 

The way it was done from #6 wasn't quite working and was crashing when you _didn't_ provide an environment variable. I've sorted that in here.

My main concern with the concept of "in source" config files is that for an end user they ideally shouldn't have to edit a source file. If they're editing this python file, it means they're running via python manually which need python installing, requirements installing etc etc. Whereas a build (like via docker) makes running easier for the end user but does take away the ability to edit source files. However this is a nice middle ground that works for both.

# Playwright
I tried modifying the build from the python-3.9 base image to the image found in #6 and the size of the image went from 89MB to 3.6GB, which seems overkill for a simple script. 

Following the docs [here](https://playwright.dev/python/docs/docker) I changed the dockerfile to properly install playwright and that resulted in 4.8GB. So ended up with the base image changing. _Looks like the build uploaded to dockerhub gets compressed, its just under 1GB so not bad_

Built the image and ran the test you suggested, led to this screenshot so looks like it works.

<img width="416" alt="image" src="https://github.com/user-attachments/assets/3b3566b4-c558-4278-a968-a16100e37c94" />

## Image Config
I've added support to have the image name come from the repo settings so that I can push to mine to test and not have to swap owner names. You'll need to add a variable to your repository settings

Navigate to:
Settings → Secrets and variables → Actions → Variables
Add the following repository variable:

DOCKER_IMAGE: eelmafia/octopus-minmax-bot

<img width="790" alt="image" src="https://github.com/user-attachments/assets/26c81f5f-f147-41a0-9dcb-591b9c4f3db5" />

joeShuff commented:
> Btw, since you mentioned your mini is acting up, have you tried keeping it next to the elec smart meter? My mini had issues when I first got it and the fix for me was to put it on the floor right under the meter itself.

My meter and home mini are about 3 meters line of sight away from each other. My meter is "in my living room" thats an extension so its tucked away in a cupboard. I'll give it a try moving closer but I've already reached out to octopus for help, I _think_ its registered to my old meters as we had an upgrade from SMETS1 to SMETS2 semi-recently.

eelmafia commented:
We should probably look into getting rid of Playwright completely because it is very chonky indeed. I used it because it was the quickest solution for me but if folk want to run this in Home Assistant etc then a more lightweight solution would be better. That's for the future though hah. 

Anyways, awesome stuff. Thanks for contributing!

joeShuff commented:
It's a new one to me, never heard of it and its definitely an interesting one!

If you want to get the building via Actions setup, you'll need to create yourself an account on dockerhub and then put the credentials into the repository secrets (make sure the secrets and not the variables). As I mentioned in the original PR comment.

Then add a variable as mentioned in the latest comment.

Then when you create a github release it will automatically build and push with the tag of the github release! If you have any questions I'd be happy to help
